src/execution_plan/ops/op_unwind.c

```diff
@@ -32,36 +37,88 @@ OpBase *NewUnwindOp(uint record_idx, AR_ExpNode **exps) {
 }
 
 OpResult UnwindInit(OpBase *opBase) {
+	OpUnwind *unwind = (OpUnwind *)opBase;
+
+	// check if there are children - if not, add static record
+	if(!unwind->op.childCount)
+		unwind->currentRecord = Record_New(opBase->record_map->record_len);
+
+	// check for modifiers in the AR_EXP for static or dynamic list
+	rax *modifies = raxNew();
+	AR_EXP_CollectEntityIDs(unwind->expressions, modifies);
+	// if there aren't any modifiers - list is static
+	if(!raxSize(modifies)) {
+		// set the list
+		unwind->list = AR_EXP_Evaluate(unwind->expressions, NULL);
+		assert(unwind->list.type == T_ARRAY);
+		unwind->isStatic = true;
+		unwind->listIdx = 0;
+	}
+
 	return OP_OK;
 }
 
+// try to generate new value to return
+// NULL will be returned if dynamic list is not evaluted (listIdx = UNDEFIND)
+// or in case of the current list fully returned its memebers
+Record _handoff(OpUnwind *op) {
+	// if there is a new value ready, return it
+	if(op->listIdx < array_len(op->list.array)) {
+		Record r = Record_Clone(op->currentRecord);
+		Record_AddScalar(r, op->unwindRecIdx, op->list.array[op->listIdx]);
+		op->listIdx++;
+		return r;
+	}
+	return NULL;
+}
+
 Record UnwindConsume(OpBase *opBase) {
 	OpUnwind *op = (OpUnwind *)opBase;
 
-	// Evaluated and returned all expressions.
-	if(op->expIdx == array_len(op->expressions)) return NULL;
+	// check for new value
+	Record res = _handoff(op);
+	if(res) return res;
+
+	// no new value - check if there are new lists to unwind
+	// check if dynamic or static unwind
+	if(op->op.childCount) {
+		Record r;
+		OpBase *child = op->op.children[0];
+		// if there are new lists to unwind
+		if((r = OpBase_Consume(child))) {
+			if(op->listIdx != UNDEFIND) Record_Free(op->currentRecord);
```

Comment:
Move this free to right before line 84.
at this point handoff didn't managed to produce data, which means we either fully consumed our array, or we don't have an array just yet.

Reply:
correct
but
I prefer to do the clean up of the last record in the `UnwindFree`, for readability reasons.
In the current state the code replaces the old record with a new record, on-demand, thus freeing the old record, and when calling `UnwindFree` the last used record is freed.